idearium-lib/lib/mq/rpc-client.js

```diff
@@ -0,0 +1,149 @@
+     * @param  {Number}   reconnectTimeout   Timeout (milliseconds) to reconnect to rabbitmq server. Defaults to 5000
+     */
+    constructor(connectionString, options, reconnectTimeout) {
+
+        if (!connectionString) {
+            throw new Error('connectionString parameter is required');
+        }
+
+        // Init Connection
+        super(connectionString, options, reconnectTimeout);
+
+        this.promises = {};
+        this.channel = undefined;
+        this.queue = undefined;
+
+        // Once we have a connection to Rabbit, connect as an RPC server.
+        this.on('connect', this.setupRpc);
+
+    }
+
+    /**
+     * An internal function to connect to RabbitMQ in RPC server mode.
+     * @return Promise A promise.
+     */
+    setupRpc() {
+
+        // Create a channel.
+        return this.connection.createChannel()
+            .then((ch) => {
+
+                // Store the channel and emit the channel connection event.
+                this.channel = ch;
+                this.emit('channel', this);
+
+                // Create a queue.
+                return ch.assertQueue('', { exclusive: true });
+
+            })
+            .then((q) => {
+
+                debug(`RPC client is awaiting publishing requests [${q.queue}]`);
+
+                // Store the queue and emit the queue connection event.
+                this.queue = q;
+                this.emit('queue', this);
+
+                return this.setupConsumer();
+
+            })
+            .catch((err) => {
+
+                if (err.message !== 'Connection closing') {
+                    throw err;
+                }
+
+            });
+
+    }
+
+    /**
+     * Used to publish a message to an RPC.
+     * @param  {String} rpcName The name of the RPC.
+     * @param  {Object} data         The data to send to the RPC.
+     * @return {Promise}             A promise that will be resolved with the result.
+     */
+    publish(rpcName, data) {
+
+        // Use the promiseManager to map promises specific to correlationIds.
+        return new Promise((resolve, reject) => this.promiseManager(rpcName, uuid.v4(), data, resolve, reject));
+
+    }
+
+    /**
+     * An internal method used to store a promise, against a correlationId, and then publish the data to the RPC.
+     * @param  {String} publishQueue  The name of the RPC.
+     * @param  {String} correlationId A v4 uuid used to correlate incoming messages to promises.
+     * @param  {Object} data          The data to send to the RPC.
+     * @param  {Function} resolve     The resolve function of a Promise.
+     * @param  {Function} reject      The reject function of a Promise.
+     * @return Void
+     */
+    promiseManager(publishQueue, correlationId, data, resolve, reject) {
+
+        // Allow 2000 milliseconds for the RPC to work.
+        const timeoutId = setTimeout(() => {
+            reject(new Error('RPC timed out.'))
+        }, 2000);
```

Comment:
I had to increase this to at lease 15000 (15 seconds) for it to work.
Considering the RPC call has to travel between internal and external systems (ones we can't fix :)), would it be worthwhile increasing this to something ridiculous to be sure the request failed? e.g. 30 seconds http://paylosophy.com/tag/credit-card-processing-timeout/